torch/_inductor/codegen/cpp.py

```diff
@@ -2817,36 +2830,76 @@ def reduction(self, dtype, src_dtype, reduction_type, value):
+                    * FloorDiv(self.ranges[self.tiling_idx], self.tiling_factor)
+                    if self.tiling_idx >= self.reduction_depth
+                    else reduction_size
+                )
+                if FloorDiv(self.ranges[self.tiling_idx], self.tiling_factor)
+                else sympy.Integer(0)
             )
-            reduction_factor = (
-                self.tiling_factor if self.tiling_idx >= self.reduction_depth else 1
+            self.welford_helper_val = self.welford_helper_cse.generate(
+                self.compute, f"reduction {reduction_key}", write=False
             )
-            self.weight_recp_vec_range = FloorDiv(reduction_size, reduction_factor)
-            if self.weight_recp_vec_range not in self.weight_recps_cse.reduction_cache:
-                self.weight_recps_val = self.weight_recps_cse.generate(
-                    self.compute, f"reduction {self.weight_recp_vec_range}", write=False
+            self.non_parallel_reduction_prefix.writeline(
+                self.welford_weight_reciprocal_vec(
+                    dtype, self.welford_helper_vec_range, self.welford_helper_val
                 )
-                self.weight_recps_cse.reduction_cache[self.weight_recp_vec_range] = (
-                    self.weight_recps_val
+            )
+            num_threads = (
+                "max_threads" if config.cpp.dynamic_threads else parallel_num_threads()
+            )
+            self.local_reduction_init.writeline(
+                self.welford_weight_reciprocal_vec(
+                    dtype,
+                    self.welford_helper_vec_range,
+                    self.welford_helper_val,
+                    num_threads,
                 )
-                self.non_parallel_reduction_prefix.writeline(
-                    self.welford_weight_reciprocal_vec(dtype)
+            )
+            self.welford_reduction_stores.writeline(
+                f"{acc_vec} = welford_combine({acc_vec}, &{self.welford_helper_val});"
+            )
+            self.local_welford_reduction_stores.writeline(
+                f"{acc_vec}_local = welford_combine({acc_vec}_local, &{self.welford_helper_val});"
+            )
+            # generate welford_helper for tail vec kernel
+            self.masked_welford_helper_vec_range = (
+                (
+                    FloorDiv(reduction_size, self.ranges[self.tiling_idx])
+                    if self.tiling_idx >= self.reduction_depth
+                    else reduction_size
                 )
-                # generate weight_recps for parallel reduction
-                num_threads = (
-                    "max_threads"
-                    if config.cpp.dynamic_threads
-                    else parallel_num_threads()
+                if self.ranges[self.tiling_idx] % self.tiling_factor
+                else sympy.Integer(0)
+            )
+            self.masked_welford_helper_val = self.masked_welford_helper_cse.generate(
```

Comment:
Can we combine the vec and mask_vec part to save code?

Reply:
Done, thanks for your comments!